src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs

```diff
@@ -10,30 +10,46 @@ internal static partial class JSHostImplementation
     {
         internal unsafe delegate void ToManagedCallback(JSMarshalerArgument* arguments_buffer);
 
-        public sealed class PromiseHolder
+        public sealed unsafe class PromiseHolder
         {
+            public bool IsDisposed;
             public readonly nint GCHandle; // could be also virtual GCVHandle
             public ToManagedCallback? Callback;
             public JSProxyContext ProxyContext;
-            public bool IsDisposed;
-            public bool IsCanceling;
 #if FEATURE_WASM_MANAGED_THREADS
             public ManualResetEventSlim? CallbackReady;
+            public PromiseHolderState* State;
 #endif
 
             public PromiseHolder(JSProxyContext targetContext)
             {
                 GCHandle = (IntPtr)InteropServices.GCHandle.Alloc(this, GCHandleType.Normal);
                 ProxyContext = targetContext;
+#if FEATURE_WASM_MANAGED_THREADS
+                State = (PromiseHolderState*)Marshal.AllocHGlobal(sizeof(PromiseHolderState));
+                Interlocked.Exchange(ref (*State).IsResolving, 0);
```

Comment:
I think this can be Volatile.Write, but I'm also not sure the difference is meaningful

Reply:
We should also explicitly memset the whole allocation to zero, because iirc AllocHGlobal isn't guaranteed to be zero-filled.

Reply:
There is just that one field in the structure. `Interlocked.Exchange` would also flush CPU cache line, I believe.